moveit_core/collision_detection_fcl/src/collision_common.cpp

```diff
@@ -439,7 +440,7 @@ bool distanceCallback(fcl::CollisionObjectd* o1, fcl::CollisionObjectd* o2, void
   }
 
   // use the collision matrix (if any) to avoid certain distance checks
-  bool always_allow_collision = false;
+  thread_local bool always_allow_collision = false;
```

Comment:
Thanks! I understand. Let's see what the results are after that.

Reply:
I updated the usage and the statistics in the PR description. The results are still great but not quite amazing. Average control period dropped from 0.0064s -> 0.0054s.

- There was an extra variable declared at L598. Deleted it.

- Question: Is it worth using `thread_local` for a const pointer? It would need to become non-const. Like this:

`const CollisionGeometryData* res_cd1 = static_cast<const CollisionGeometryData*>(fcl_result.o1->getUserData());`

Could become:

`thread_local CollisionGeometryData* cd1`
`cd1 = static_cast<const CollisionGeometryData*>(o1->collisionGeometry()->getUserData());`

Reply:
I'm on very thin ice here and am definitely not experienced in optimizing runtime like that.
I have no idea if we have anyone in the core community who is experienced enough to judge these patches from a professional background.

My main concern is that this still looks very much like don't-outsmart-your-compiler, even if you provide some data from your personal setup... I would actually be interested in a godbolt analysis and some gperf statistics to show the effect of these changes.
Turning local variables into statics forces compilers to create a symbol for them.
So whereas the local variable might have ended up in a register only, suddenly the compiler must turn them into reads/writes to memory. Even if a local double variable is stored on the stack it essentially boils down to a `sub esp,0x08`
and a `mov` assignment from the active cache line. On the other hand the thread_local memory is in a different part of the address space, so this can actually break cache lines.
<details><summary>Example from my system</summary>

<p>

```
$ cat foo.cpp ; make foo; ./foo
#include <iostream>
using namespace std;

int main(){
        int x; cout << &x << ": local" << endl;
        thread_local int z; cout << &z << ": thread_local" << endl;
        static int y; cout << &y << ": static" << endl;
        return 0;
}
g++     foo.cpp   -o foo
0x7ffc4dbc031c: local
0x7f5673dfdf7c: thread_local
0x404198: static
```
</p>
</details>

If anything, I would expect things like [the hot attribute](https://github.com/TAMS-Group/bio_ik/blob/master/src/ik_evolution_2.cpp#L242) and [likely/unlikely markers](https://clang.llvm.org/docs/AttributeReference.html#likely-and-unlikely) might somewhat improve behavior.

> Is it worth using thread_local for a const pointer? It would need to become non-const.

Definitely not! Const-correctness is more important than somewhat dubious optimizations.

Reply:
Please see my comment on new benchmarking methodology below. Better?